### PR TITLE
feat(device-service): CAN device service

### PR DESCRIPTION
**Pull request by vijay-hcl-epl**
Adding initial version of CAN device service.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+DOCKERS=docker_device_can_c
+.PHONY: $(DOCKERS)
+
+VERSION=$(shell cat ./VERSION || echo 0.0.0)
+GIT_SHA=$(shell git rev-parse HEAD)
+
+build: ./VERSION ${MICROSERVICES}
+
+build/release/device-can-c/device-can-c:
```
Comment (lenny-intel) [inline Makefile:15]:
```suggestion
build/release/device-can/device-can:
```

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,39 @@
+
+.PHONY: build test clean docker
+
+MICROSERVICES=build/release/device-can-c/device-can-c
+.PHONY: $(MICROSERVICES)
+
+DOCKERS=docker_device_can_c
```
Comment (lenny-intel) [inline Makefile:7]:
```suggestion
DOCKERS=docker_device_can
```

File: OWNERS.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Repository Owners
+
+This repository is managed by the EdgeX Device Services Working Group.  As such, the **Device Services Working Group** chairman is considered the "owner" of the repository and approves all committers of the repository.
+
+See the [project Wiki TSC page](https://wiki.edgexfoundry.org/pages/viewpage.action?pageId=329436#TechnicalSteeringCommittee(TSC)-WorkingGroups) for information on the current EdgeX TSC and who occupies the role of Device Service Working Group chair.
+
+For a complete list of current committers see:  https://github.com/orgs/edgexfoundry/teams/device-bacnet-c-committers/members.
```
Comment (lenny-intel) [inline OWNERS.md:7]:
```suggestion
For a complete list of current committers see:  https://github.com/orgs/edgexfoundry/teams/device-can-committers/members.
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,109 @@
...
+- Linux has support for CAN drivers
+- Character device-based drivers – can4linux
+- Network socket-based drivers - socketcan
+- Each supports different controllers and protocols
+- This link(https://elinux.org/CAN_Bus) shows the different controllers and protocols supported by can4linux & socketcan
+
+## CAN Device Service Overview
+- Socketcan is used
+- The link - https://docs.kernel.org/networking/can.html#motivation-why-using-the-socket-api mentions why socketcan is preferred over character device-based driver
+- C – device-sdk-c is used. Version - 3.0.2
```
Comment (lenny-intel) [inline README.md:32]: Remove version as it quickly becomes stall
```suggestion
- C – device-sdk-c is used. 
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,109 @@
...
+candump -tz can0 (or vcan0)
+
+### Run the Service
+
+Ensure other EdgeX services are already running.
+
+Once the build is completed and interfaces are enabled, run the executable with the below commands
+
+export EDGEX_SECURITY_SECRET_STORE=false
+./build/release/device-can-c
```
Comment (lenny-intel) [inline README.md:84]:
```suggestion
./build/release/device-can
```

File: scripts/Dockerfile.alpine [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+# limitations under the License.
+#
+
+ARG BASE=alpine:3.16
+FROM ${BASE} as builder
+RUN wget https://iotech.jfrog.io/artifactory/api/security/keypair/public/repositories/alpine-release -O /etc/apk/keys/alpine.dev.rsa.pub
+RUN echo 'https://iotech.jfrog.io/artifactory/alpine-release/v3.16/main' >> /etc/apk/repositories
+RUN apk add --update --no-cache build-base git gcc cmake make linux-headers yaml-dev libmicrohttpd-dev curl-dev util-linux-dev ncurses-dev paho-mqtt-c-dev-1.3 hiredis-dev libcbor-dev iotech-iot-1.5-dev
+
+COPY scripts /device-can-c/scripts
```
Comment (lenny-intel) [inline scripts/Dockerfile.alpine:23]: Change all occurrences of  `device-can-c` to `device-can`

File: scripts/Dockerfile.alpine [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+COPY --from=builder /usr/lib/libcsdk.so /usr/lib
+COPY --from=builder /usr/share/doc/edgex-csdk /usr/share/doc/edgex-csdk
+
+COPY LICENSE /.
+COPY Attribution.txt /.
+COPY res /res
+COPY sample-profiles /sample-profiles
+
+LABEL license='SPDX-License-Identifier: Apache-2.0' \
+      copyright='Copyright (c) 2019-2022: IoTech Ltd'
```
Comment (lenny-intel) [inline scripts/Dockerfile.alpine:49]: Change copyright to HCL

File: scripts/Dockerfile.alpine-base [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+# limitations under the License.
+#
+
+ARG BASE=alpine:3.16
+FROM ${BASE}
+RUN wget https://iotech.jfrog.io/artifactory/api/security/keypair/public/repositories/alpine-release -O /etc/apk/keys/alpine.dev.rsa.pub
+RUN echo 'https://iotech.jfrog.io/artifactory/alpine-release/v3.16/main' >> /etc/apk/repositories
+RUN apk add --update --no-cache build-base git gcc cmake make linux-headers yaml-dev libmicrohttpd-dev curl-dev util-linux-dev ncurses-dev paho-mqtt-c-dev-1.3 hiredis-dev libcbor-dev iotech-iot-1.5-dev
+
+COPY scripts /device-can-c/scripts
```
Comment (lenny-intel) [inline scripts/Dockerfile.alpine-base:23]: Change all occurrences of  `device-can-c` to `device-can`

**Review by lenny-intel — CHANGES_REQUESTED** [review]
I reviewed structure, build files, etc. Will leave C code to the C experts.

**Comment by lindseysimple** [comment]
@vijay-hcl-epl Since this repo was not created from the template-repo, please  add a `.github` directory (see device-bacnet-c [repository](https://github.com/edgexfoundry/device-bacnet-c/tree/main/.github)) to align with EdgeX standards per @MightyNerdEric  's comment, thanks.

File: Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,21 @@
+//
+// Copyright (c) 2019 Intel Corporation
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply Jenkinsfile]: Corrected

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,39 @@
+
+.PHONY: build test clean docker
+
+MICROSERVICES=build/release/device-can-c/device-can-c
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply Makefile]: Corrected

File: Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+edgeXBuildCApp (
+    project: 'device-can-c',
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply Jenkinsfile]: Corrected

File: Makefile (hunk 1 above)
Comment (vijay-hcl-epl) [reply Makefile]: Corrected

File: Makefile (hunk 2 above)
Comment (vijay-hcl-epl) [reply Makefile]: Corrected

File: OWNERS.md (hunk 3 above)
Comment (vijay-hcl-epl) [reply OWNERS.md]: Corrected

File: README.md (hunk 5 above)
Comment (vijay-hcl-epl) [reply README.md]: Corrected

File: scripts/Dockerfile.alpine [hunk 12]
```diff
@@ -0,0 +1,53 @@
+#
+# Copyright (c) 2020 IOTech Ltd
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply scripts/Dockerfile.alpine]: Corrected

File: scripts/Dockerfile.alpine (hunk 6 above)
Comment (vijay-hcl-epl) [reply scripts/Dockerfile.alpine]: Corrected

File: scripts/Dockerfile.alpine (hunk 7 above)
Comment (vijay-hcl-epl) [reply scripts/Dockerfile.alpine]: Corrected

File: scripts/Dockerfile.alpine-base [hunk 13]
```diff
@@ -0,0 +1,31 @@
+#
+# Copyright (c) 2020 IOTech Ltd
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply scripts/Dockerfile.alpine-base]: Corrected

File: scripts/Dockerfile.alpine-base (hunk 8 above)
Comment (vijay-hcl-epl) [reply scripts/Dockerfile.alpine-base]: Corrected

**Comment by vijay-hcl-epl** [comment]
@lenny-intel , @lindseysimple  Thank you for the review comments. I have corrected all the changes and added a new commit. 

Please let me know whether do I need to add the file Attribution.txt?

Thank You,
Vijay

**Review by lenny-intel — COMMENTED** [review]
LGTM form structure POV will leave config and C code to the C device service experts.

**Comment by lindseysimple** [comment]
> @lenny-intel , @lindseysimple Thank you for the review comments. I have corrected all the changes and added a new commit.
> 
> Please let me know whether do I need to add the file Attribution.txt?
> 
> Thank You, Vijay

Thanks @vijay-hcl-epl. Yes, we'll need to add Attribution.txt for the CAN device service as well. Please see [device-coap-c](https://github.com/edgexfoundry/device-coap-c/blob/main/Attribution.txt) for a reference.

**Comment by lenny-intel** [comment]
> Please see [device-coap-c](https://github.com/edgexfoundry/device-coap-c/blob/main/Attribution.txt) for a reference.

@lindseysimple , the current C devices don't have proper Attribution.tx files. I have submitted issues to fix them.

@vijay-hcl-epl , please use https://github.com/edgexfoundry/device-mqtt-go/blob/main/Attribution.txt as a sample for the format. It should include all dependencies including the one from the C SDK.

**Comment by lenny-intel** [comment]
@vijay-hcl-epl , I have reviewed the changes for all my comments and marked them as resolved. It is now up to the C experts to review the code.

File: src/c/device-can.c [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+
+	iot_log_debug(impl->lc, "CAN device service started..\n");
+	while (!quit);
+
+	devsdk_service_stop(service, true, &e);
+	ERR_CHECK(e);
+
+	devsdk_service_free(service);
+	iot_data_free(driver_map);
+	free(impl);
```
Comment (FelixTing) [inline src/c/device-can.c:372]: add `free(canImpls;)` here

File: src/c/device-can.h [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+ 1. Sathya Durai           HCL Technologies Ltd (EPL Team)
+ 2. Vijay Annamalaisamy    HCL Technologies Ltd (EPL Team)
+*/
+
+#ifndef _DEVICE_CAN_H_
+#define _DEVICE_CAN_H_ 1
+
+/**
+ * @file
+ * @brief Defines common artifacts for the CoAP device service.
```
Comment (FelixTing) [inline src/c/device-can.h:21]:
```suggestion
 * @brief Defines common artifacts for the CAN device service.
```

File: scripts/Dockerfile.alpine [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+RUN apk add --update --no-cache linux-headers yaml libmicrohttpd curl libuuid paho-mqtt-c-dev-1.3 hiredis libcbor iotech-iot-1.5
+
+COPY --from=builder /device-can/build/release/ /
+COPY --from=builder /usr/lib/libcsdk.so /usr/lib
+COPY --from=builder /usr/share/doc/edgex-csdk /usr/share/doc/edgex-csdk
+
+COPY LICENSE /.
+COPY Attribution.txt /.
+COPY res /res
+COPY sample-profiles /sample-profiles
```
Comment (FelixTing) [inline scripts/Dockerfile.alpine:46]: The `sample-profiles` directory is not present.

File: scripts/Dockerfile.alpine [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+RUN echo 'https://iotech.jfrog.io/artifactory/alpine-release/v3.16/main' >> /etc/apk/repositories
+RUN apk add --update --no-cache build-base git gcc cmake make linux-headers yaml-dev libmicrohttpd-dev curl-dev util-linux-dev ncurses-dev paho-mqtt-c-dev-1.3 hiredis-dev libcbor-dev iotech-iot-1.5-dev
+
+COPY scripts /device-can/scripts
+COPY src /device-can/src/
+COPY VERSION /device-can/
+
+WORKDIR /device-can
+
+RUN /device-can/scripts/build_deps.sh
```
Comment (FelixTing) [inline scripts/Dockerfile.alpine:29]:
```suggestion
RUN /device-can/scripts/build_deps.sh 1
```

File: src/c/device-can.c [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+		return x.code;                                        \
+	}
+
+#define CAN_INTFC_ADDR "CanInterface"
+#define FILTER_MSG_ID1 "FilterMsgId1"
+#define FILTER_MASK1   "FilterMask1"
+#define TIMEOUT "Timeout"
+
+can_driver *impl;
+iot_data_t *can_resp_data;
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply src/c/device-can.c:55]: removed can_resp_data

File: src/c/device-can.c (hunk 14 above)
Comment (vijay-hcl-epl) [reply src/c/device-can.c:372]: added free(canImpls)

File: src/c/device-can.h (hunk 15 above)
Comment (vijay-hcl-epl) [reply src/c/device-can.h:21]: Corrected

File: scripts/Dockerfile.alpine (hunk 17 above)
Comment (vijay-hcl-epl) [reply scripts/Dockerfile.alpine:29]: modified

File: src/c/device-can.c [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+===============================================================
+ 1. Sathya Durai           HCL Technologies Ltd (EPL Team)
+ 2. Vijay Annamalaisamy    HCL Technologies Ltd (EPL Team)
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply src/c/device-can.c:20]: removed unused headers

File: src/c/device-can.c [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+#include <sys/time.h>
+#include <sys/types.h>
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply src/c/device-can.c:22]: removed unused headers

File: src/c/device-can.c [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+#include <signal.h>
+
+#include <linux/can.h>
+#include <linux/can/raw.h>
+
+#include <iot/data.h>
+#include "device-can.h"
+
+#include <stdarg.h>
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (vijay-hcl-epl) [reply src/c/device-can.c:37]: removed unused headers

**Comment by vijay-hcl-epl** [comment]
> > Please see [device-coap-c](https://github.com/edgexfoundry/device-coap-c/blob/main/Attribution.txt) for a reference.
> 
> @lindseysimple , the current C devices don't have proper Attribution.txt files. I have submitted issues to fix them.
> 
> @vijay-hcl-epl , please use https://github.com/edgexfoundry/device-mqtt-go/blob/main/Attribution.txt as a sample for the format. It should include all dependencies including the one from the C SDK.

Thank you @lenny-intel & @lindseysimple. I have made the changes and submitted a commit

**Comment by vijay-hcl-epl** [comment]
Thank you @FelixTing. I have made the changes and submitted a new commit.

**Review by lindseysimple — APPROVED** [review]
The project structure, `.github` directory content and `Addtribution.txt` all looks good to me, thanks Vijay.

**Review by FelixTing — APPROVED** [review]
LGTM. Will leave the device callback implementation, which is specific to the CANbus protocol, to CANbus specialists.

File: Attribution.txt [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+
+edgexfoundry/go-mod-secrets (Apache 2.0) https://github.com/edgexfoundry/go-mod-secrets/v3
+https://github.com/edgexfoundry/go-mod-secrets/blob/master/LICENSE
+
+edgexfoundry/go-mod-configuration (Apache 2.0) https://github.com/edgexfoundry/go-mod-configuration/v3
+https://github.com/edgexfoundry/go-mod-configuration/blob/master/LICENSE
+
+edgexfoundry/go-mod-messaging (Apache 2.0) https://github.com/edgexfoundry/go-mod-messaging/v3
+https://github.com/edgexfoundry/go-mod-messaging/blob/master/LICENSE
+
```
Comment (lenny-intel) [inline Attribution.txt:29]: These are all got Go, not C. Please remove

File: scripts/Dockerfile.alpine [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+COPY VERSION /device-can/
+
+WORKDIR /device-can
+
+RUN /device-can/scripts/build_deps.sh 1
+
+RUN /device-can/scripts/build.sh 
+
+FROM alpine:3.16
+MAINTAINER iotech <support@iotechsys.com>
```
Comment (lenny-intel) [inline scripts/Dockerfile.alpine:34]: PLease update the Maintainer to be HCL

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: Attribution.txt (hunk 22 above)
Comment (vijay-hcl-epl) [reply Attribution.txt:29]: Removed

File: scripts/Dockerfile.alpine (hunk 23 above)
Comment (vijay-hcl-epl) [reply scripts/Dockerfile.alpine:34]: Updated

**Review by lenny-intel — APPROVED** [review]
LGTM

**Comment by lenny-intel** [comment]
@lindseysimple , @FelixTing , I think we are ready for TSC vote to accept this new device service into EdgeX.

**Comment by lindseysimple** [comment]
> @lindseysimple , @FelixTing , I think we are ready for TSC vote to accept this new device service into EdgeX.

Thanks @lenny-intel. I'll send a email for TCS vote.

**Pull request merged**